engines/sci/graphics/gfxdrivers.cpp

```diff
@@ -0,0 +1,508 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "common/events.h"
+#include "common/system.h"
+
+#include "graphics/cursorman.h"
+#include "graphics/paletteman.h"
+
+#include "sci/graphics/gfxdrivers.h"
+#include "sci/resource/resource.h"
+#include "sci/sci.h"
+#include "sci/version.h"
+
+namespace Sci {
+
+Common::Point GfxDriver::getMousePos() const {
+	return g_system->getEventManager()->getMousePos();
+}
+
+GfxDefaultDriver::GfxDefaultDriver(uint16 screenWidth, uint16 screenHeight) : GfxDriver(screenWidth, screenHeight, 0, 1) {
+	switch (g_sci->getResMan()->getViewType()) {
+	case kViewEga:
+		_numColors = 16;	// QFG PC-98 with 8 colors also reports 16 here
+		break;
+	case kViewAmiga:
+		_numColors = 32;
+		break;
+	case kViewAmiga64:
+		_numColors = 64;
+		break;
+	case kViewVga:
+	case kViewVga11:
+		_numColors = 256;
+		break;
+	default:
+		break;
+	}
+
+	if (_numColors == 0)
+		error("GfxDefaultDriver: Unknown view type");
+}
+
+void GfxDefaultDriver::setPalette(const byte *colors, uint start, uint num) {
+	g_system->getPaletteManager()->setPalette(colors, start, num);
+}
+
+void GfxDefaultDriver::copyRectToScreen(const byte *src, int pitch, int x, int y, int w, int h) {
+	g_system->copyRectToScreen(src, pitch, x, y, w, h);
+}
+
+void GfxDefaultDriver::replaceCursor(const void *cursor, uint w, uint h, int hotspotX, int hotspotY, uint32 keycolor) {
+	CursorMan.replaceCursor(cursor, w, h, hotspotX, hotspotY, keycolor);
+}
+
+SCI0_DOSPreVGADriver::SCI0_DOSPreVGADriver(int numColors, int screenW, int screenH, int horizontalAlignment) : GfxDriver(screenW, screenH, numColors, horizontalAlignment), _palNeedUpdate(true), _colors(nullptr), _compositeBuffer(nullptr) {
+	_compositeBuffer = new byte[screenW * screenH]();
+}
+
+SCI0_DOSPreVGADriver::~SCI0_DOSPreVGADriver() {
+	delete[] _compositeBuffer;
+}
+
+bool SCI0_DOSPreVGADriver::checkDriver(const char *const *driverNames, int listSize) {
+	Common::String missing;
+	while (listSize-- && *driverNames) {
+		if (Common::File::exists(*driverNames))
+			return true;
+		if (!missing.empty())
+			missing += " or ";
+		missing += "'" + Common::String(*driverNames) + "'";
+		++driverNames;
+	}
+	warning("Driver file %s not found. Starting game in EGA mode", missing.c_str());
+	return false;
+}
+
+void SCI0_DOSPreVGADriver::assignPalette(const byte *colors) {
+	_colors = colors;
+}
+
+void SCI0_DOSPreVGADriver::setPalette(const byte*, uint, uint) {
+	if (!_palNeedUpdate || !_colors)
+		return;
+	_palNeedUpdate = false;
+	byte *tmp = new byte[768]();
+	memcpy(tmp, _colors, _numColors * 3);
+	g_system->getPaletteManager()->setPalette(tmp, 0, 256);
+	delete[]tmp;
+}
+
+SCI0_EGADriver::SCI0_EGADriver() : SCI0_DOSPreVGADriver(16, 320, 200, 1) {
+	static const uint8 egaColors[] = {
```

Comment:
I'm assuming that the other 16 colour video drivers use the same palette?

Reply:
The EGA driver implementation could actually be ditched completely. I made this class structure at the beginning, when I still thought that the dithering etc. would happen inside the driver. But the EGA driver really does nothing to modify any of the incoming graphics. Same as the MCGA driver (which I haven't implemented), it really just sends the data int the hardware. So that could really be cleaned up...